File: src/station.py
#!/usr/bin/env python3

"""Implementation of the Python server for the CITS3002 project.

Author: Eddie Atkinson (22487668)
Date: 06/05/2020

This module contains the main function for the program and is responsible for parsing the
command line arguments passed to it, initialising a node object,
opening the appropriate TCP and UDP ports, sending name frames to its
neighbours, and finally entering the select loop that is used to process
requests

Typical usage example:
./station.py <station-name> <tcp-port> <udp-port> <neighbour-1-udp-port> ...

Code linted using mypy
"""
import socket
import sys
import os
import select
from time import sleep
from Frame import Frame
from FrameType import FrameType
from constants import HOST
from constants import MAX_PACKET_LEN
from udp import process_udp
from tcp import process_tcp
from Node import Node


def main() -> None:
    """Starting point for the program

    Parses the command line args, initialises a node object representing this station,
    sends name frames to neighbours and then enters the select loop
    """
    this_node = parse_args()
    print(f"Node: {this_node.name} running with PID {os.getpid()}")
    try:
        init_ports(this_node)
        this_node.check_kill()
        send_name_frames(this_node)
        listen_on_ports(this_node)
    except OSError:
        print("Address most likely in use")
        this_node.quit(1)


def init_ports(this_node: Node) -> None:
    """Initialises the UDP and TCP ports for an node.

    Args:
        this_node: An instance of the node class representing a given node in the network.

    Returns:
        None
    """
    tcp_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    tcp_socket.bind((HOST, this_node.tcp_port))
    tcp_socket.listen(5)
    tcp_socket.setblocking(False)

    udp_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    udp_socket.bind((HOST, this_node.udp_port))
    udp_socket.setblocking(False)

    # Wait for everyone else to bind their ports
    sleep(5)

    this_node.tcp_socket = tcp_socket
    this_node.udp_socket = udp_socket
    print("You can send frames now")


def send_name_frames(this_node: Node) -> None:
    """Constructs a frame containing a node's name and send it to its neighbours.

    Args:
        this_node: An instance of the node class representing a given node in the network.
    Returns:
        None
    """
    name_frame = Frame(this_node.name, "", [this_node.name], -1, -1, FrameType.NAME_FRAME)
    for port in this_node.neighbours.keys():
        this_node.udp_socket.sendto(name_frame.to_bytes(), (HOST, port))


def listen_on_ports(this_node: Node) -> None:
    """Starts the node's select loop, handling both UDP and TCP packets.

    Args:
        this_node: An instance of the node class representing a given node in the network.
    Returns:
        None
    """
    this_node.input_sockets = [this_node.tcp_socket, this_node.udp_socket]
    while True:
        this_node.check_kill()
        readers = select.select(this_node.input_sockets, [], [], 10.0)[0]
        for reader in readers:
            if reader is this_node.udp_socket:
                transmission = reader.recvfrom(MAX_PACKET_LEN)
                process_udp(this_node, transmission)
            elif reader is this_node.tcp_socket:
                #  We have a new tcp connection
                conn, addr = reader.accept()
                print(f"New connection from {addr}")
                conn.setblocking(False)
                this_node.input_sockets.append(conn)
            else:
                transmission = reader.recv(MAX_PACKET_LEN)
                if not transmission:
                    # Disconnection
                    print(f"{reader} has disconnected")
                    this_node.input_sockets.remove(reader)
                    reader.shutdown(socket.SHUT_RD)
                    reader.close()
                else:
                    process_tcp(this_node, transmission, reader)


def parse_args() -> Node:
    """Parses the arguments passed to the program and returns an initialised node object.

    Args:
        None
    Returns:
        An instance of the Node class containing a node's name, TCP port, UDP port,
        and the UDP ports of its neighbours.
    """
    args = list(sys.argv)
    # Don't really care about the program name
    args.pop(0)

    name = args.pop(0)
    tcp_port = int(args.pop(0))
    udp_port = int(args.pop(0))

    neighbours = {}
    # The rest of the arguments should be our neighbours ports
    while args:
        neighbours[int(args.pop(0))] = ""

    this_node = Node(name, neighbours, tcp_port, udp_port)
    return this_node


if __name__ == "__main__":
    main()

File: src/station.cpp
#include "station.h"

Node parse_args(int argc, char *argv[]) {
  Node this_node = Node();
  this_node.name = argv[1];
  this_node.tcp_port = (uint16_t)atoi(argv[2]);
  this_node.udp_port = (uint16_t)atoi(argv[3]);

  for (int i = 4; i < argc; ++i) {
    uint16_t port = (uint16_t)atoi(argv[i]);
    this_node.neighbours.insert({port, ""});
  }
  return this_node;
}

void send_name_frames(Node &this_node) {
  list<string> new_src;
  Frame name_frame = Frame(this_node.name, "", new_src, -1, -1, NAME_FRAME);
  string frame_str = name_frame.to_string();
  size_t len = frame_str.size();
  struct sockaddr_in out_addr;
  out_addr.sin_family = AF_INET;
  inet_pton(AF_INET, HOST_IP, &(out_addr.sin_addr));

  for (auto neighbour : this_node.neighbours) {
    uint16_t port = neighbour.first;
    out_addr.sin_port = htons(port);
    if (sendto(this_node.udp_socket, frame_str.c_str(), len, 0,
               (struct sockaddr *)&out_addr, sizeof out_addr) < 0) {
      cout << this_node.name << " error in sending frame " << frame_str << endl;
      this_node.quit(1);
    }
  }
  return;
}

void handle_sockets(Node &this_node, fd_set *rfds) {
  char buf[MAX_PACKET_LEN];
  size_t len = MAX_PACKET_LEN;
  struct sockaddr_in from;
  socklen_t fromlen = sizeof(from);
  bzero(&from, fromlen);
  ssize_t read;
  for (unsigned int i = 0; i < this_node.input_sockets.size(); ++i) {
    int socket = this_node.input_sockets.at(i);
    if (FD_ISSET(socket, rfds)) {
      if (socket == this_node.udp_socket) {
        // UDP received
        read = recvfrom(this_node.udp_socket, &buf, len, 0,
                        (struct sockaddr *)&from, &fromlen);
        string frame(buf, read);
        uint16_t port = ntohs(from.sin_port);
        process_udp(this_node, frame, port);
      } else if (socket == this_node.tcp_socket) {
        // New TCP connection
        int new_sock =
            accept(this_node.tcp_socket, (struct sockaddr *)&from, &fromlen);
        if (fcntl(new_sock, F_SETFL, O_NONBLOCK) != 0) {
          cout << "New socket failed to set non-blocking, exiting" << endl;
          this_node.quit(1);
        }
        this_node.input_sockets.push_back(new_sock);
        cout << "Received new connection from " << inet_ntoa(from.sin_addr)
             << ":" << ntohs(from.sin_port) << endl;
      } else {
        // Servicing existing TCP connections
        size_t tcp_read = 0;
        tcp_read = recv(socket, buf, len, 0);
        if (tcp_read == 0) {
          // Connection has closed
          this_node.remove_socket(socket);
          cout << "Closing connection to " << socket << endl;
        } else if (tcp_read > 0) {
          string frame(buf, read);
          handle_tcp(this_node, frame, socket);
        } else {
          cout << "Failed read operation on TCP socket, exiting" << endl;
          this_node.quit(1);
        }
      }
    }
  }
}

void listen_on_ports(Node &this_node) {
  struct timeval tv;
  fd_set read_fds;
  int fdsready;
  int largestfd;
  while (true) {
    this_node.check_kill();
    FD_ZERO(&read_fds);
    tv.tv_sec = 5;
    tv.tv_usec = 0;
    fdsready = 0;
    largestfd = -1;
    for (auto socket : this_node.input_sockets) {
      if (socket > largestfd) {
        largestfd = socket;
      }
      FD_SET(socket, &read_fds);
    }
    fdsready = select(largestfd + 1, &read_fds, NULL, NULL, &tv);
    switch (fdsready) {
    case (-1):
      cout << "Error with select, exiting" << endl;
      this_node.quit(1);
      break;
    case (0):
      break;
    default:
      handle_sockets(this_node, &read_fds);
      break;
    }
  }
}

int main(int argc, char *argv[]) {
  if (argc <= 2) {
    cout << "Server needs arguments to work" << endl;
    exit(1);
  }
  Node this_node = parse_args(argc, argv);
  cout << "Node: " << this_node.name << " running with PID " << getpid()
       << endl;
  this_node.init_ports();
  send_name_frames(this_node);
  listen_on_ports(this_node);
  return 0;
}
